### ISSUE TITLE
Uncaught exception when fusing sparse and dense dimensions

### ISSUE THREAD

**Issue by remysucre** [tensor-compiler/taco#527]
It probably does not make sense to fuse a sparse dimension with a dense one, but currently this is unchecked: 

```
remywang@m16 build % ./bin/taco "q(i,j,k) = r(i,j) * s(j,k) * t(i,k)" -f=q:sss -f=r:sd -f=s:ss -f=t:ss -s="fuse(i,j,f)"                             
libc++abi: terminating with uncaught exception of type taco::TacoException: Compiler bug at /Users/remywang/projects/taco/src/lower/lowerer_impl_imperative.cpp:873 in lowerForall
Please report it to developers
 Condition failed: underivedAncestors.size() == 1
 
zsh: abort      ./bin/taco "q(i,j,k) = r(i,j) * s(j,k) * t(i,k)" -f=q:sss -f=r:sd -f=s:ss  
```

**Comment by rohany** [comment]
Yes, I believe fusing without iterating over the position space is supported for only dense dimensions currently. I'm not sure what the best way of fusing co-iteration loops looks like.

**Comment by rohany** [comment]
Take a look at http://tensor-compiler.org/senanayake-oopsla20-taco-scheduling.pdf, section 2 (and later in the paper) for some details on `pos`. The key idea of pos is to enable iteration over the non-zero values of a tensor only, and apply further transformations such as strip-mining to this space of non-zeros.

The scheduling system doesn't seem to throw an error (it doesn't do the best pre-condition checking), but your schedule is invalid. Here's a valid schedule doing maybe what you intended:

```
bin/taco "q = r(i,j) * s(i,j)" -f=r:ss -s="fuse(i, j, f)", -s="pos(f, fpos, r)"
```

which first fuses `i` and `j` into `f`, and then declares `f` to range over the position space of `r` -- over the non-zeros of `r`. This is useful then to strip-mine `fpos`, which generates statically load-balanced iterations over the non-zeros of `r`.

**Comment by remysucre** [comment]
I see, thanks! I'll leave this open for the uncaught errors, but feel free to close if those are not of high priority.
